juju/juju
# Add support for serving the new dashboard

## Description of change

Add support for serving the new dashboard GUI.
The old GUI endpoint is retained as well for upgraded controllers.

The juju dashboard CLI will prefer to look for the new /dashboard endpoint but will fall back to using the /gui endpoint for older controllers or older controllers that have been upgraded.

Support is added for the new metadata "min-juju-version" attribute. This allows newer dashboards to be released that might not be compatible with older Juju versions.

Also add a new model config attribute to allow (when bootstrapping) a devel version of the dashboard to be used. Previously Juju was hard coded to only look for release metadata.
The upgrade-gui command has also been changed to support --gui-stream

The new dashboard tarball changes how content is packed. There's no nested folders and there's a separate version file.

## QA steps

Check that a dashboard can be set up manually via JUJU_GUI.

`export JUJU_GUI=jujugui-2.16.0.tar.bz2`

then bootstrap and run `juju dashboard`
It should print out the dashboard URL and credential and going to that in a browser should server the new dashboard.

Also, try using the published beta dashboard

`juju bootstrap --config gui-stream=devel lxd`

To test older GUI, bootstrap a 2.7 controller.
Running `juju dashboard` should print the older GUI URL.

Upgrade the controller to 2.8
Running `juju dashboard` should print the older GUI URL.
Also then check that a browser can use the older GUI from the upgraded controller

Now upgrade to the new dashboard
`juju upgrade-dashboard jujugui-2.16.0.tar.bz2`

Run `juju dashboard` and it will print the new dashboard URL and test in a browser.
